.github/workflows/heroku.yml

```diff
@@ -0,0 +1,23 @@
+name: deploy to heroku
+on:
+  push:
+    paths-ignore:
+      - 'README.md'
+  pull_request:
+    paths-ignore:
+      - 'README.md'
+jobs:
+ heroku:
+    if: github.event_name == 'push' && github.ref == 'refs/heads/exercise8'
+    
+    name: heroku deploy
+    runs-on: ubuntu-latest
+    steps:
+    - name: Check out repository
+      uses: actions/checkout@v2
+    - name: Deploy to Heroku
+      uses: akhileshns/heroku-deploy@v3.12.12
+      with:
+          heroku_api_key: ${{ secrets.HEROKU_API_KEY }}
+          heroku_app_name: ${{ secrets.HEROKU_APP_NAME }}
+          heroku_email: ${{ secrets.HEROKU_EMAIL }}
```

Comment:
I think the main issue here is that we haven't told Heroku to try deploying with Docker:
```suggestion
          heroku_email: ${{ secrets.HEROKU_EMAIL }}
          usedocker: true
```

The addition I'd add is that this particular action unfortunately doesn't have an easy option (if I remember rightly!) for setting the `target` for our Docker build, so it just uses the default stage which is the final one defined in your Dockerfile. To get it working, we will likely want to rearrange your Dockerfile slightly so that instead of:
```
FROM base as production
CMD <prod command>

FROM base as development
ENTRYPOINT <dev command>

FROM base as test
ENTRYPOINT <test command>
```

We have:
```
FROM base as development
ENTRYPOINT <dev command>

FROM base as test
ENTRYPOINT <test command>

# production should be the last stage to allow the CD pipeline to build the prod image by default
FROM base as production
CMD <prod command>
```